extensions/MozChangeField/web/js/set-tracking-severity-s1.js

```diff
@@ -0,0 +1,112 @@
+ *
+ * This Source Code Form is "Incompatible With Secondary Licenses", as
+ * defined by the Mozilla Public License, v. 2.0. */
+
+/**
+ * Reference or define the Bugzilla app namespace.
+ * @namespace
+ */
+var Bugzilla = Bugzilla || {}; // eslint-disable-line no-var
+
+/**
+ * Enforce setting of Priority to P1 when Severity is set to S1
+ */
+Bugzilla.SetTrackingSeverityS1 = class SetTrackingSeverityS1 {
+  /**
+   * Initialize a new SetTrackingSeverityS1 instance.
+   */
+  constructor() {
+    this.priority = document.querySelector("#priority");
+    this.severity = document.querySelector("#bug_severity");
+    this.flags = document.querySelector("div.edit-show table.tracking-flags");
+
+    if (this.severity && this.priority && this.flags) {
+      this.sev_curr_value = this.severity.value;
+      //this.severity.addEventListener("change", () => this.severity_onselect());
+
+      // Find cf_tracking_ specific flags and
+      // store current values to reset them if needed
+      this.fetchProductDetails().then((data) => {
+        const product_details = this.getMajors(data);
+        const nightly = "cf_tracking_firefox" + product_details.nightly;
+        const beta = "cf_tracking_firefox" + product_details.beta;
+        this.flag_selects = [];
+        this.flag_curr_values = [];
+        this.flag_curr_titles = [];
+        this.flags
+          .querySelectorAll(
+            'select[name="' + nightly + '"], select[name="' + beta + '"]'
+          )
+          .forEach((flag) => {
+            this.flag_selects.push(flag);
+            this.flag_curr_values[flag.name] = flag.value;
+            this.flag_curr_titles[flag.name] = flag.title;
+          });
+      });
+    }
+  }
+
+  /**
+   * Called when severity select is changed. If severity if changed to S1
+   * from something else, set firefox beta and nightly tracking flags to '?'.
+   */
+  severity_onselect() {
+    const s1_selected =
+      this.sev_curr_value !== "S1" && this.severity.value === "S1";
+
+    this.flag_selects.forEach((flag) => {
+      const options = flag.querySelectorAll("option");
+      options.forEach((opt) => (opt.disabled = s1_selected));
+      if (s1_selected && flag.value !== "+") {
+        const request_opt = Array.from(options).filter(
+          (opt) => opt.value === "?"
+        )[0];
+        request_opt.disabled = false;
+        request_opt.selected = true;
+        flag.title = "Flag is locked since Severity was moved to S1";
+      } else if (
+        flag.value !== this.flag_curr_values[flag.name] &&
+        flag.value === "?"
+      ) {
+        const request_opt = Array.from(options).filter(
+          (opt) => opt.value === this.flag_curr_values[flag.name]
+        )[0];
+        request_opt.disabled = false;
+        request_opt.selected = true;
+        flag.title = this.flag_curr_titles[flag.name];
+      }
+    });
+  }
+
+  // Below borrowed from extensions/BMO/web/js/firefox-crash-table.js
+  async fetchProductDetails() {
```

Comment:
we already have this information cached server-side; pass it into the page rather than re-fetching on every page load.

Reply:
Yep. I first check for the existence of the meta[name="firefox-versions"] element which has the JSON already populated for show_bug.cgi. If that is not present, we then try to load it ourselves.

Reply:
We _must not_ duplicate the logic here.  If the data isn't available then fallback to sane defaults rather than attempting to fetch here.

Reply:
Sorry for the confusion here. I can definitely move this code from here and firefox-crash-table.js to the util.js file to remove duplication. But what would be sane defaults be in this case other than just returning an empty JSON object? I could not just return some arbitrary version numbers.

Reply:
This code needs to be removed; the data should _always_ come from the server.  If the perl code is unable to fetch the product details then attempting to do it client-side will only DDoS the server.

If product-details is down the best course of action here is to do nothing – not set tracking-? flags.